File: lib/javascript/src/stop.ts
import {
	assertInRange,
	assertLength,
	readUint16,
	writeUint16,
} from "./bytes.ts";
import { TRAM_ROUTE } from "./route.ts";

/**
 * Where a transaction happened. Urban bus stops carry an internal stop id that is scoped to the
 * route rather than shared across the network, tram stops their stop number, and any other
 * operator an id of its own; on Cercanías the low byte looks like the station's position along
 * the line, counting from the far terminus.
 */
export type Stop =
	| { network: "urban"; id: number }
	| { network: "tram"; stop: number }
	| { network: "other"; id: number };

const URBAN_FLAG = 0x8000;

/**
 * Decodes bytes 5 and 6 of a transaction.
 * @param bytes The two bytes.
 * @param route Route id of the transaction, needed to tell the tram from other operators.
 */
export function decodeStop(bytes: Uint8Array, route: number): Stop {
	assertLength(bytes, 2, "stop");
	const value = readUint16(bytes, 0);
	if (value & URBAN_FLAG) return { network: "urban", id: value & ~URBAN_FLAG };
	if (route === TRAM_ROUTE) return { network: "tram", stop: value };
	return { network: "other", id: value };
}

/** Encodes a stop into two bytes. */
export function encodeStop(stop: Stop): Uint8Array {
	const bytes = new Uint8Array(2);
	switch (stop.network) {
		case "urban":
			assertInRange("urban stop id", stop.id, 0, URBAN_FLAG - 1);
			writeUint16(bytes, 0, URBAN_FLAG | stop.id);
			break;
		case "tram":
			assertInRange("tram stop", stop.stop, 0, URBAN_FLAG - 1);
			writeUint16(bytes, 0, stop.stop);
			break;
		case "other":
			assertInRange("stop id", stop.id, 0, URBAN_FLAG - 1);
			writeUint16(bytes, 0, stop.id);
	}
	return bytes;
}

File: lib/rust/src/stop.rs
use crate::bytes::in_range;
use crate::error::Result;
use crate::route::Route;

/// Where a transaction happened, bytes 5 and 6 of a transaction.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Stop {
	/// Bit 15 set: an urban bus stop with an internal stop id.
	Urban(u16),
	/// Bit 15 clear on the tram route: the stop number × 100.
	Tram(u16),
	/// Bit 15 clear on any other route: an id of that operator.
	Other(u16),
}

const URBAN_FLAG: u16 = 0x8000;
const MAX_ID: u32 = 0x7fff;

impl Stop {
	/// Decodes two big-endian bytes; `route` tells the tram from other operators.
	#[must_use]
	pub fn decode(bytes: [u8; 2], route: Route) -> Self {
		let value = u16::from_be_bytes(bytes);
		if value & URBAN_FLAG != 0 {
			Self::Urban(value & !URBAN_FLAG)
		} else if route == Route::TRAM {
			Self::Tram(value)
		} else {
			Self::Other(value)
		}
	}

	/// Encodes into two big-endian bytes.
	///
	/// # Errors
	/// [`Error::Range`](crate::Error::Range) above `0x7FFF`.
	pub fn encode(self) -> Result<[u8; 2]> {
		let value = match self {
			Self::Urban(id) => URBAN_FLAG | in_range("urban stop id", id, 0, MAX_ID)?,
			Self::Tram(stop) => in_range("tram stop", stop, 0, MAX_ID)?,
			Self::Other(id) => in_range("stop id", id, 0, MAX_ID)?,
		};
		Ok(value.to_be_bytes())
	}
}
